<br><br><br><br><br>

# Getting ROOT data into Numpy

<br><br><br><br><br>

<br><br><br><br><br><br>

<span style="font-size: 24px; margin-left: 230px"><b>Method 1: PyROOT</b></span>

<br><br><br><br><br><br>

In [1]:
import ROOT

file = ROOT.TFile("data/HZZ.root")
tree = file.Get("events")

for i, event in enumerate(tree):
    print(event.MET_px, event.MET_py)
    if i > 10:
        break

Welcome to JupyROOT 6.14/04
5.912771224975586 2.5636332035064697
24.76520347595215 -16.349109649658203
-25.78508758544922 16.237131118774414
8.619895935058594 -22.78654670715332
5.393138885498047 -1.3100523948669434
-3.7594752311706543 -19.417020797729492
23.962148666381836 -9.049156188964844
-57.533348083496094 -20.48767852783203
42.416194915771484 -94.35086059570312
-1.9144694805145264 -23.96303367614746
19.710058212280273 4.645508766174316
-35.538055419921875 -14.753822326660156


In [2]:
# That worked, but it would be too slow for large datasets.
# PyROOT does A LOT of processing each time it is invoked, so we don't want that in a loop over events.

# However, there's an alternative that goes straight to Numpy:
tree.AsMatrix(["MET_px", "MET_py"])

array([[  5.91277122,   2.5636332 ],
       [ 24.76520348, -16.34910965],
       [-25.78508759,  16.23713112],
       ...,
       [ 18.10164642,  50.29071808],
       [ 79.87519073, -52.35145187],
       [ 19.71374893,  -3.59541821]])

In [3]:
# If you need the branch names, you can get them with return_labels=True

data, labels = tree.AsMatrix(["MET_px", "MET_py"], return_labels=True)
print(f"labels = {labels}\n")

import pandas
pandas.DataFrame(data, columns=labels)

labels = ['MET_px', 'MET_py']



MET_px      MET_py
0       5.912771    2.563633
1      24.765203  -16.349110
2     -25.785088   16.237131
3       8.619896  -22.786547
4       5.393139   -1.310052
5      -3.759475  -19.417021
6      23.962149   -9.049156
7     -57.533348  -20.487679
8      42.416195  -94.350861
9      -1.914469  -23.963034
10     19.710058    4.645509
11    -35.538055  -14.753822
12     15.209903   -6.397500
13     28.459160    5.657520
14     -9.093978    0.861486
15      0.350357   -4.255576
16     -0.201901   -3.499514
17      8.286861   -3.266164
18     -7.835592  -13.773303
19    -41.880852 -157.988144
20     -1.715726   19.486519
21      2.373739   -8.089888
22    -30.234993    2.027190
23     13.711685   -6.066514
24    130.723969   74.998268
25     -6.254472   18.144054
26    -14.468836   16.849669
27     -1.018478   -8.156774
28     12.017437    1.811257
29    -14.349060   14.992454
...          ...         ...
2391    5.421937  -40.910957
2392   10.979458    4.336320
2393   34.796410   -2.099867
2394    0.745414  -54.461945
2395   19.757299  -44.224178
2396  104.612236  -98.111153
2397    3.185825   79.226997
2398    1.150845   -3.101060
2399    4.515370   -4.102074
2400  -10.098569  -54.974846
2401    2.998099   21.993523
2402   27.944883  -37.503429
2403    3.787466  -21.217653
2404    9.378232   27.572329
2405  -17.310106    5.082886
2406  -81.965927  -11.167821
2407   -9.059591  -24.464552
2408   25.649775   -2.818718
2409   29.691553   56.606071
2410  -25.754967   14.037041
2411   -2.426847  -30.494087
2412  -15.611773   -9.571323
2413   18.921183    0.019401
2414  -11.730723   11.629858
2415  -10.648725  -20.464689
2416  -14.607650  -28.204895
2417   22.208313   59.774940
2418   18.101646   50.290718
2419   79.875191  -52.351452
2420   19.713749   -3.595418

[2421 rows x 2 columns]

In [4]:
# If your data have more than one value per event, PyROOT can handle that...

for i, event in enumerate(tree):
    print(len(event.Muon_Px), "muons:", [x for x in event.Muon_Px])
    if i > 10:
        break

2 muons: [-52.89945602416992, 37.7377815246582]
1 muons: [-0.8164593577384949]
2 muons: [48.987831115722656, 0.8275666832923889]
2 muons: [22.08833122253418, 76.6919174194336]
2 muons: [45.171321868896484, 39.75095748901367]
2 muons: [9.228110313415527, -5.793715000152588]
2 muons: [12.538717269897461, 29.541839599609375]
1 muons: [34.883758544921875]
2 muons: [-53.16697311401367, 11.491869926452637]
2 muons: [-67.01485443115234, -18.118755340576172]
2 muons: [15.983028411865234, 34.68440628051758]
2 muons: [-70.51190948486328, -38.028743743896484]


In [5]:
# ...but tree.AsMatrix cannot.

tree.AsMatrix(["Muon_Px"])

array([[0.],
       [0.],
       [0.],
       ...,
       [0.],
       [0.],
       [0.]])

Error in <TTreeReaderValueBase::GetBranchDataType()>: Must use TTreeReaderArray to read branch Muon_Px: it contains an array or a collection.
Error in <TTreeReaderValueBase::CreateProxy()>: The branch Muon_Px contains data of type {UNDETERMINED TYPE}, which does not have a dictionary.


<img src="img/03-coming-soon-1.png" width="93%">

<img src="img/03-coming-soon-2.png" width="93%">

<br><br><br><br><br><br>

<span style="font-size: 24px; margin-left: 230px"><b>Method 2: root_numpy</b></span>

<br><br><br><br><br><br>

In [6]:
# root_numpy is built on top of C++ ROOT, not PyROOT
import root_numpy

# it can extract Numpy arrays from a PyROOT tree
root_numpy.tree2array(tree, ["MET_px", "MET_py"])

# or directly from a filename/treename
root_numpy.root2array("data/HZZ.root", "events", ["MET_px", "MET_py"])

array([(  5.912771,   2.5636332), ( 24.765203, -16.34911  ),
       (-25.785088,  16.237131 ), ..., ( 18.101646,  50.290718 ),
       ( 79.87519 , -52.35145  ), ( 19.713749,  -3.5954182)],
      dtype=[('MET_px', '<f4'), ('MET_py', '<f4')])

And it uses all the same tricks as `TTree::Draw` to loop over events with minimal overhead.

<img src="img/root-numpy-fast.png" width="100%">

In [7]:
# Unlike PyROOT's AsMatrix, root_numpy can read multiple values per event.

# Each event is represented as a separate Numpy array, and
# they're all in a Numpy object array (Numpy array of pointers to Python objects).

root_numpy.tree2array(tree, ["Muon_Px"])

array([(array([-52.899456,  37.73778 ], dtype=float32),),
       (array([-0.81645936], dtype=float32),),
       (array([48.98783  ,  0.8275667], dtype=float32),), ...,
       (array([-29.756786], dtype=float32),),
       (array([1.1418698], dtype=float32),),
       (array([23.913206], dtype=float32),)], dtype=[('Muon_Px', 'O')])

<br>

**Remember what a Numpy array is: an array per event means a Python object per event.**

<br>

<center><img src="img/numpy-memory-layout.png" width="80%"></center>

<br>

In [8]:
# Worse still, the pointers (array.ctypes.data) are not contiguous in memory: pointer chasing!

[x[0].ctypes.data for x in root_numpy.tree2array(tree, ["Muon_Px"])[:20]]

[106830502020656,
 106830502022192,
 106830502020720,
 106830488856816,
 106830502020784,
 106830463478496,
 106830507367344,
 106830502020688,
 106830487525728,
 106830487525792,
 106830494344672,
 106830494344736,
 106830502020816,
 106830494344832,
 106830494344896,
 106830488488752,
 106830488488816,
 106830488488880,
 106830488488944,
 106830508000000]

In [9]:
# Furthermore, it's not a single array, so multidimensional indexing doesn't work.

a = root_numpy.tree2array(tree, ["Muon_Px"])

a[:10, 0]

IndexError: too many indices for array

<img src="img/root-numpy-optionalroot.png" width="100%">

In [10]:
# But the advantage of being compiled against ROOT is that you can run any TTree::Draw formula
# on the branches, including C++ object methods.

# This muonp4 branch has type TLorentzVector, which has a Pt() method.
pandas.DataFrame(
    root_numpy.root2array("data/HZZ-objects.root", "events", ["muonp4.Pt()", "muonp4.Pt() + 1000"]))

muonp4.Pt()  \
0                [54.16810703140204, 37.74415265978988]   
1                                   [24.41791248135961]   
2               [53.58826697278532, 29.811997139120674]   
3                [88.63194310332618, 77.95148447265454]   
4                [81.01140452083426, 47.17504574667714]   
5              [41.591052358050575, 30.844215085173435]   
6               [44.35777431661225, 29.874452374029385]   
7                                   [38.37087544930784]   
8              [106.28355631699014, 12.311635885539262]   
9               [85.53881159059787, 39.506103898783294]   
10               [51.64914852288953, 46.49716868294783]   
11              [75.51093887887036, 50.514888290524446]   
12                                  [79.79030751726177]   
13                                 [26.225671131477768]   
14             [128.02862618591368, 49.620239052606124]   
15               [54.80017239533791, 43.27432148588847]   
16               [53.99401375649458, 31.03810019611473]   
17                                  [46.81564562809829]   
18              [48.38432826266856, 22.844403427258605]   
19    [50.34849799613474, 33.98432269384266, 19.5256...   
20             [59.241336101266675, 31.045482083886156]   
21              [39.38447172785867, 29.775330796393686]   
22              [51.413430422821015, 40.14969736673915]   
23                                  [55.47143806913377]   
24                                 [62.739702675727685]   
25             [51.941238403544965, 35.264491830259054]   
26                                  [25.80016715106026]   
27                                   [50.3112282694008]   
28              [70.71846468347898, 28.587417576868287]   
29              [71.20280452402069, 47.464213230610625]   
...                                                 ...   
2391                                [30.04504545372089]   
2392                                 [70.9793623592288]   
2393            [90.21392820819183, 40.946725336532985]   
2394            [27.89672504255123, 27.801072886784407]   
2395                               [56.903531066383415]   
2396            [107.60963327057247, 40.88883377098777]   
2397             [68.9053871546623, 27.247716276377126]   
2398                                [24.18790064023774]   
2399             [52.86494512751539, 19.32115221145384]   
2400            [59.978915429217956, 56.37943127741561]   
2401                                [35.36808761769148]   
2402                               [39.741086188290936]   
2403                               [31.330797436088073]   
2404                                [63.80593339751661]   
2405            [52.443837271808086, 36.12020344001634]   
2406            [76.18733843936784, 31.307274249720876]   
2407           [45.434622183033774, 29.265601602084903]   
2408                                                 []   
2409                                [32.09086668290633]   
2410               [58.388242188928, 35.83115445691242]   
2411            [61.64505369183525, 28.647494884161652]   
2412                                [35.27358198566274]   
2413                               [42.554395216040724]   
2414                                [43.19601861306685]   
2415             [44.97159626392337, 33.24442978298967]   
2416                                [41.91365807511886]   
2417                                [37.81439514806715]   
2418                               [33.461536523629654]   
2419                               [63.619817705512375]   
2420                                [42.93994828083644]   

                                     muonp4.Pt() + 1000  
0                 [1054.168107031402, 1037.74415265979]  
1                                  [1024.4179124813595]  
2              [1053.5882669727853, 1029.8119971391206]  
3              [1088.6319431033262, 1077.9514844726546]  
4              [1081.0114045208343, 1047.1750457466771]  
5              [1041.5910523580505, 1

<br><br><br><br><br><br>

<span style="font-size: 24px; margin-left: 230px"><b>Method 3: uproot</b></span>

<br><br><br><br><br><br>

<img src="img/uproot-frontpage.png" width="100%">

In [11]:
import uproot

# uproot uses the Python standard library and Numpy to read ROOT files into
# 
#     dict of bytestring → array

uproot.open("data/HZZ.root")["events"].arrays(["MET_px", "MET_py"])

{b'MET_px': array([  5.912771,  24.765203, -25.785088, ...,  18.101646,  79.87519 ,
         19.713749], dtype=float32),
 b'MET_py': array([  2.5636332, -16.34911  ,  16.237131 , ...,  50.290718 ,
        -52.35145  ,  -3.5954182], dtype=float32)}

In [12]:
# A variable number of values per event is represented as a jagged array:

uproot_tree = uproot.open("data/HZZ.root")["events"]
jagged = uproot_tree.array("Muon_Px")
jagged

<JaggedArray [[-52.899456 37.73778] [-0.81645936] [48.98783 0.8275667] ... [-29.756786] [1.1418698] [23.913206]] at 0x70e84eb1b048>

In [13]:
# These per-event arrays are contiguous in memory and NOT Python objects.

# An array for a single event (e.g. when presenting it on the screen) is generated on demand.

print(f"\njagged presentation: {jagged}")
print(f"\njagged content: {jagged.content}")
print(f"\njagged offsets: {jagged.offsets}")


jagged presentation: [[-52.899456 37.73778] [-0.81645936] [48.98783 0.8275667] ... [-29.756786] [1.1418698] [23.913206]]

jagged content: [-52.899456    37.73778     -0.81645936 ... -29.756786     1.1418698
  23.913206  ]

jagged offsets: [   0    2    3 ... 3823 3824 3825]


In [30]:
%%timeit

for event in tree:
    [x for x in event.Muon_Px]

19.9 ms ± 1.3 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


In [31]:
%%timeit

root_numpy.tree2array(tree, ["Muon_Px"])

1.51 ms ± 23.3 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


In [32]:
%%timeit

uproot_tree.array("Muon_Px")

743 µs ± 15.9 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


With minor caveats, this is how data are stored in the bytes of the ROOT file on disk, so

    PyROOT/root_numpy: columnar → event objects → columnar

               uproot: columnar         →         columnar

| Method     | Time (ms) | Speedup |
|:-----------|----------:|--------:|
| PyROOT     | 19.9      | 1×      |
| root_numpy | 1.51      | 13×     |
| uproot     | 0.743     | 27×     |

(Note: this is a tiny file, running time dominated by starting and stopping.)

**Here's an analysis of a more substantial file.**

<br>

<img src="img/logscales.png" width="100%">

In [15]:
# Being a Python reimplementation, class instances do not have any of their C++ methods.

# But many class types can be read (because ROOT includes streamers in the file to decode them.)

muons = uproot.open("data/HZZ-objects.root")["events"].array("muonp4")
muons

<JaggedArrayMethods [[TLorentzVector(-52.899, -11.655, -8.1608, 54.779) TLorentzVector(37.738, 0.69347, -11.308, 39.402)] [TLorentzVector(-0.81646, -24.404, 20.2, 31.69)] [TLorentzVector(48.988, -21.723, 11.168, 54.74) TLorentzVector(0.82757, 29.801, 36.965, 47.489)] ... [TLorentzVector(-29.757, -15.304, -52.664, 62.395)] [TLorentzVector(1.1419, 63.61, 162.18, 174.21)] [TLorentzVector(23.913, -35.665, 54.719, 69.556)]] at 0x70e84ecf3208>

In [33]:
# And a few classes have methods implemented in Python.

print("muons.pt =", muons.pt)
print("\nmuons[:10].mass =", muons[:10].mass)

leading    = muons[muons.counts >= 2, 0]
subleading = muons[muons.counts >= 2, 1]
print("\n(leading + subleading) =", leading + subleading)
print("\n(leading + subleading).mass =", (leading + subleading).mass)

muons.pt = [[54.16810703140204 37.74415265978988] [24.41791248135961] [53.58826697278532 29.811997139120674] ... [33.461536523629654] [63.619817705512375] [42.93994828083644]]

muons[:10].mass = [[0.10559298741436905 0.10545247041042287] [0.105499240400313] [0.10696309110601164 0.10513788128369116] ... [0.1177042005466644] [0.10003178932782261 0.10570681544295826] [0.10358060768576076 0.10761322229556605]]

(leading + subleading) = [TLorentzVector(-15.162, -10.961, -19.468, 94.181) TLorentzVector(49.815, 8.0774, 48.133, 102.23) TLorentzVector(98.78, -99.792, 738.94, 757.5) ... TLorentzVector(22.797, -5.2167, 32.613, 100.43) TLorentzVector(28.806, 16.557, -24.978, 95.018) TLorentzVector(2.9387, 18.416, -261.92, 273.35)]

(leading + subleading).mass = [90.22779777 74.74654928 89.75736376 ... 92.06495256 85.44384208
 75.96066262]


<br><br><br><br><br>

# Basic uproot

<br><br><br><br><br>

In [36]:
# We've already seen some examples; now let's break them down to see how they work.

file = uproot.open("http://scikit-hep.org/uproot/examples/Event.root")

# A file or TDirectory object acts like a Python dict.
print(file.keys(), end="\n\n")

# The value is not read from disk until you ask (and every time you ask).
print(file.values(), end="\n\n")

# You can check the classes before opening it.
print(file.classes(), end="\n\n")

[b'ProcessID0;1', b'htime;1', b'T;1', b'hstat;1']

[<TProcessID b'ProcessID0' at 0x70e861e58518>, <b'TH1F' b'htime' 0x70e861e79f98>, <TTree b'T' at 0x70e861e4c860>, <b'TH1F' b'hstat' 0x70e861de0b88>]

[(b'ProcessID0;1', <class 'uproot.rootio.TProcessID'>), (b'htime;1', <class 'uproot.rootio.TH1F'>), (b'T;1', <class 'uproot.rootio.TTree'>), (b'hstat;1', <class 'uproot.rootio.TH1F'>)]



In [39]:
# Exercise: extract a histogram from "file" and call .show() on it to view its data.

# How would you do that if it were a Python dict?

                   0                                                     0.38739
                   +-----------------------------------------------------------+
[-inf, 0) 0.021839 |***                                                        |
[0, 1)    0.33352  |***************************************************        |
[1, 2)    0.30403  |**********************************************             |
[2, 3)    0.32452  |*************************************************          |
[3, 4)    0.35097  |*****************************************************      |
[4, 5)    0.36894  |********************************************************   |
[5, 6)    0.30728  |***********************************************            |
[6, 7)    0.30681  |***********************************************            |
[7, 8)    0.34156  |****************************************************       |
[8, 9)    0.16151  |*************************                                  |
[9, 10)   0        |        

In [44]:
# nested directories may be navigated with / for convenience.
file = uproot.open("http://scikit-hep.org/uproot/examples/nesteddirs.root")

file["one"]["two"]["tree"].show()
# file["one/two/tree"].show()

Int32                      (no streamer)              asdtype('>i4')
Int64                      (no streamer)              asdtype('>i8')
UInt32                     (no streamer)              asdtype('>u4')
UInt64                     (no streamer)              asdtype('>u8')
Float32                    (no streamer)              asdtype('>f4')
Float64                    (no streamer)              asdtype('>f8')
Str                        (no streamer)              asstring()
ArrayInt32                 (no streamer)              asdtype("('>i4', (10,))")
ArrayInt64                 (no streamer)              asdtype("('>i8', (10,))")
ArrayUInt32                (no streamer)              asdtype("('>u4', (10,))")
ArrayUInt64                (no streamer)              asdtype("('>u8', (10,))")
ArrayFloat32               (no streamer)              asdtype("('>f4', (10,))")
ArrayFloat64               (no streamer)              asdtype("('>f8', (10,))")
N                          (no streamer) 

In [47]:
tree = file["one/two/tree"]

# TTrees are like dicts for their branches, which have an array() method.
# You can also call array() or arrays() on the TTree itself.

print(tree["Int32"].array())
print(tree.array("Int32"))
print(tree.arrays(["Int32", "Float32"]))

[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23
 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47
 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 68 69 70 71
 72 73 74 75 76 77 78 79 80 81 82 83 84 85 86 87 88 89 90 91 92 93 94 95
 96 97 98 99]
[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23
 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47
 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 68 69 70 71
 72 73 74 75 76 77 78 79 80 81 82 83 84 85 86 87 88 89 90 91 92 93 94 95
 96 97 98 99]
{b'Int32': array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
       34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50,
       51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67,
       68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84,
  

In [51]:
# Array branches become n-dimensional Numpy arrays (len(shape) > 1)

print(tree["ArrayInt32"].title)
tree.array("ArrayInt32")

b'ArrayInt32[10]/I'


array([[ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 1,  1,  1,  1,  1,  1,  1,  1,  1,  1],
       [ 2,  2,  2,  2,  2,  2,  2,  2,  2,  2],
       [ 3,  3,  3,  3,  3,  3,  3,  3,  3,  3],
       [ 4,  4,  4,  4,  4,  4,  4,  4,  4,  4],
       [ 5,  5,  5,  5,  5,  5,  5,  5,  5,  5],
       [ 6,  6,  6,  6,  6,  6,  6,  6,  6,  6],
       [ 7,  7,  7,  7,  7,  7,  7,  7,  7,  7],
       [ 8,  8,  8,  8,  8,  8,  8,  8,  8,  8],
       [ 9,  9,  9,  9,  9,  9,  9,  9,  9,  9],
       [10, 10, 10, 10, 10, 10, 10, 10, 10, 10],
       [11, 11, 11, 11, 11, 11, 11, 11, 11, 11],
       [12, 12, 12, 12, 12, 12, 12, 12, 12, 12],
       [13, 13, 13, 13, 13, 13, 13, 13, 13, 13],
       [14, 14, 14, 14, 14, 14, 14, 14, 14, 14],
       [15, 15, 15, 15, 15, 15, 15, 15, 15, 15],
       [16, 16, 16, 16, 16, 16, 16, 16, 16, 16],
       [17, 17, 17, 17, 17, 17, 17, 17, 17, 17],
       [18, 18, 18, 18, 18, 18, 18, 18, 18, 18],
       [19, 19, 19, 19, 19, 19, 19, 19, 19, 19],
       [20, 20, 20, 

In [52]:
# And arbitrary-length array branches become JaggedArrays.

print(tree["SliceInt32"].title)
tree.array("SliceInt32")

b'SliceInt32[N]/I'


<JaggedArray [[] [1] [2 2] ... [97 97 97 ... 97 97 97] [98 98 98 ... 98 98 98] [99 99 99 ... 99 99 99]] at 0x70e84e3289b0>